Unify cmdlets with parameter 'Encoding' to be of type System.Text.Encoding

JamesWTruher opened the pull request:
This unifies file encoding across the inbox cmdlets to be UTF-8 without a BOM for all platforms. This is a breaking change as cmdlets on windows have a number of different encodings. This  supports better interoperability with tradition Linux shells as we are using the same encoding.

Use `ArgumentCompletionsAttribute` for the parameter to accept the currently used strings.
Register encoding providers on all platforms. This seems safe enough, and there are defintely more encodings available on CoreFX after this is done, which will be useful for some of the web cmdlet scenarios

Validate that files are created with UTF-8 encoding without BOM
Update tests to validate Encoding parameter to new type and create new tests for
parameter type validation.

*Breaking Change* The `-Encoding Byte` has been removed from the filesystem provider cmdlets. A new parameter -Byte is now the signal that a byte stream is required as input or output will be a stream of bytes.

JamesWTruher committed:
Unify cmdlets with parameter 'Encoding' to be of type System.Text.Encoding

Created support classes so tab completion would continue to wor
Create an ArgumentTransformationAttribute for the parameter to accept the currently used strings
Register encoding providers on all platforms. This seems safe enough, and there are defintely more
encodings available on CoreFX after this is done, which will be useful for some of the web cmdlet scenarios

Validate that files are created with UTF-8 encoding without BOM
Update tests to validate Encoding parameter to new type and create new tests for
parameter type validation.

adityapatwardhan commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1210,77 +1211,26 @@ internal static FileSystemCmdletProviderEncoding GetEncoding(string path)
-        } // GetEncodingFromEnum
 
+        // BigEndianUTF32 encoding is possible, but requires creation
+        internal static Encoding BigEndianUTF32Encoding = new UTF32Encoding(true,true);
```
Use named parameters instead of `true, true`

JamesWTruher replied:
fixed

adityapatwardhan commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7658,27 +7613,15 @@ public class FileSystemContentDynamicParametersBase
-            {
-                return Encoding == FileSystemCmdletProviderEncoding.Byte;
-            } // get
+            get { return _usingByteEncoding; }
```
Please use automatic property and remove _usingByteEncoding.

JamesWTruher replied:
fixed

adityapatwardhan commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7688,9 +7631,14 @@ public bool WasStreamTypeSpecified
             {
-                return (Encoding != FileSystemCmdletProviderEncoding.String);
-            } // get
+                return _wasStreamTypeSpecified;
```
Same as above.

JamesWTruher replied:
fixed

adityapatwardhan commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -109,16 +104,17 @@ internal static Encoding GetDefaultEncoding()
 
         /// <summary>
         /// Facade for getting OEM encoding
+        /// OEM encodings work on all platforms, or rather codepage 437 is available on both Windows 
```
Probably the comment intended to be ` codepage 437 is available on both Windows and Non-Windows`

JamesWTruher replied:
indeed so, fixed

adityapatwardhan commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -515,5 +542,104 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
+        internal class ByteEncoding : System.Text.Encoding
+        {
+            // the encoding for this is not bigendian and not BOM
+            public Encoding ActualEncoding = new UnicodeEncoding(false, false);
```
Use named parameters

JamesWTruher replied:
fixed

adityapatwardhan commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -515,5 +542,104 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
+                {
+                    // the encodingName is guaranteed to be valid, so it is safe to pass null to method
+                    // Convert(Cmdlet cmdlet, string encoding) as the value of 'cmdlet'.
+                    return EncodingConversion.Convert(null, encodingName);
```
use named parameter.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Core/TestGetCommand.Tests.ps1:
```diff
@@ -211,7 +211,7 @@
         $paramName = "Encoding"
         $results = get-command -verb get -noun content -Encoding Unicode
```
Now that you have touched the file. Can you fix the casing for the Get-Command cmdlets. 
get-command -> Get-Command 
-verb -> -Verb
-noun -> -Noun

JamesWTruher replied:
done

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+
+    Context "ParameterType for parameter 'Encoding' should be 'Encoding'" {
+        BeforeAll {
+            $testCases = get-command -module Microsoft.PowerShell.* | 
```
Please fix Get-Command casing

JamesWTruher replied:
done

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        AfterEach { 
+            if ( test-path $outputFile ) {
```
Casing for Test-Path, Remove-Item -Force

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        It "Export-CSV creates file with UTF-8 encoding without BOM" {
+            [pscustomobject]@{ Key = $testStr } | export-csv $outputFile
```
Casing Export-Csv

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+        }
+
+        It "Export-CliXml creates file with UTF-8 encoding without BOM" {
+            [pscustomobject]@{ Key = $testStr } | export-clixml $outputFile
```
Casing Export-CliXml

JamesWTruher replied:
fixed

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            $Expected = $( $ExpectedWithNewline; $ExpectedWithNewline )
+            $bytes -join "-" | should be ($Expected -join "-")
+        }
+
```
Extra line?

JamesWTruher replied:
removed

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            param ( $Command, $parameters, $Expected, $Operator)
+            & $command @parameters
+            $bytes = Get-FileBytes $outputFile
+            $bytes -join "-" | should ${Operator} ($Expected -join "-")
```
If the operator is always `Be` we can directly use it here and remove it from `$simpleTestCases`

adityapatwardhan replied:
Should this be `BeExactly`?

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            [pscustomobject]@{ Key = $testStr } | export-csv $outputFile
+            $bytes = Get-FileBytes $outputFile
+            $bytes[0,1,2] -join "-" | should not be ($utf8Preamble -join "-")
+            $bytes -join "-" | should match ($utf8bytes -join "-")
```
Should this be `MatchExactly`

JamesWTruher replied:
not needed (as they're numbers) - `MatchExactly` isn't the same as `Be`, it's just a case-sensitive match `-cmatch` I just need to be sure that my string is in the output somewhere

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            [pscustomobject]@{ Key = $testStr } | export-clixml $outputFile
+            $bytes = Get-FileBytes $outputFile
+            $bytes[0,1,2] -join "-" | should not be ($utf8Preamble -join "-")
+            $bytes -join "-" | should match ($utf8bytes -join "-")
```
Should this be `MatchExactly`

JamesWTruher replied:
not needed - for the same reason as above, i'm just making sure that the magic bytes are in the file somewhere

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            ${testStr} >> $outputFile
+            $bytes = Get-FileBytes $outputFile
+            $Expected = $( $ExpectedWithNewline; $ExpectedWithNewline )
+            $bytes -join "-" | should be ($Expected -join "-")
```
Should this be `BeExactly`

JamesWTruher replied:
nope - they're numbers so case doesn't provide value

adityapatwardhan commented on test/powershell/engine/Basic/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+        }
+        It "Encoding parameter of command '<Command>' is type 'Encoding'" -testcase $testCases {
+            param ( $command )
+            $command.Parameters['Encoding'].ParameterType.FullName | Should Be "System.Text.Encoding"
```
Should be `BeExactly`?

JamesWTruher replied:
it could be - `BeExactly` is a case-sensitive string compare, but the case of the response is not the point, only that it is the proper fullname.
I'm in the file so I'll update it, but we should be cautious about being overly pedantic.

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs:
```diff
@@ -523,4 +523,4 @@ public override object Transform(EngineIntrinsics engineIntrinsics, object input
     }
 
     #endregion
-}
```
Missed adding the change for the Encoding parameter?

dantraMSFT commented:
>     internal static class EncodingConversion

I think this class should be in its own CS file; it's not a path utility and discoverability is problematic. 

---
Refers to: src/System.Management.Automation/utils/PathUtils.cs:446 in 6ca1339. [](commit_id = 6ca13399470454bdc9a7087e2f1d95b494bd3531, deletion_comment = False)

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +206,27 @@ public SwitchParameter NoClobber
+            }
+            set
+            {
+                if ( value == EncodingConversion.byteEncoding )
```
I believe you want to use Object.ReferenceEquals or override Equals in your ByteEncoding class and handle equality there; otherwise, you're relying on Encoding.Equals which may not do what you expect.

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +206,27 @@ public SwitchParameter NoClobber
+            }
+            set
+            {
+                if ( value == EncodingConversion.byteEncoding )
```
These set methods need a comment as to why the comparison to EncodingConversion.byteEncoding is needed to provide a guide when creating new cmdlets that provide an Encoding parameter.

lzybkr replied:
Maybe add a helper, something like:
```c#
EncodingConversion.SetEncodingParameterHelper(ref _encoding, value);
```
Or:
```c#
_encoding = EncodingConversion.EncodingParameterHelper(value);
```


---
In reply to: [144073559](https://github.com/PowerShell/PowerShell/pull/5080#discussion_r144073559) [](ancestors = 144073559)

dantraMSFT commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -515,5 +542,104 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
+        // We need a way to provide an encoding and also
+        // a way to notify that the user wants "byte" encoding
+        // which is just a stream
+        internal class ByteEncoding : System.Text.Encoding
```
Should this class and the EncodingConversion.byteEncoding moved here as a public field?

In other words, is the problem you're solving with this custom instance be something third-party cmdlets will need to solve as well?

JamesWTruher replied:
this is really a dodge for backward compatibility. `-Encoding Byte` should have never existed, but it does and I don't want to have this be a difference between 6 and 5.x. This _byte_ encoding is really needed for the *-content cmdlets where the parameter is used as a flag indicating that the user is looking for a stream of bytes, I would rather not make this public

SteveL-MSFT replied:
Since we're already making this big breaking change, I think we should fix this by not having `ByteEncoding` as `Byte` is an artificial "encoding".  The `*-Content` cmdlets should have a different parameter to output or receive a byte stream.

iSazonov replied:
It would more clear to users to have -Byte or -Binary parameter.
@mklement0 @rkeithhill What do you think?

SteveL-MSFT replied:
I kind of prefer `-AsByteArray` myself.

cc @HemantMahawar @joeyaiello

mklement0 replied:
Purely from a usability perspective, I don't have a problem with `-Encoding Byte`, because even though it is technically not a _text_ encoding, the `-Encoding` parameter can be conceptualized as being about how to _interpret_ the data on reading and writing, and `-Encoding Byte` (a byte stream) is one way of (non-)interpreting .

This keeps all considerations regarding encoding/non-encoding focused on a single parameter, which simplifies things (it doesn't introduce another parameter that you have to learn about and remember), both in terms of use and documentation.

iSazonov replied:
How can we interpret `Set-Content -Encoding 'Ru-ru' -Byte`?

iSazonov commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -93,14 +93,9 @@ internal static Encoding GetDefaultEncoding()
-                uint currentAnsiCp = NativeMethods.GetACP();
-                s_defaultEncoding = Encoding.GetEncoding((int)currentAnsiCp);
-#endif
+                s_defaultEncoding = new UTF8Encoding(false);
```
I don't like the change. Previously (discussion #3248, PR #3467) I fixed this for Windows because we're still very sensitive to the OEM encoding. We have tons applications in enterprises that doesn't work if we don't set the correct OEM encoding on the system. If MSFT wants to cut it off, do it first in Windows 10 kernel and get feedback.

JamesWTruher replied:
@iSazonov where are tests which provide validation for the changes you made in PR3467? The discussion in #3248 doesn't seem to have any concrete examples of misbehavior, which I would really like to understand. As for your last, I would rather get feedback on breaking behavior outside of the Windows ship cycles.

SteveL-MSFT replied:
@iSazonov, @PowerShell/powershell-committee has already decided to go with UTF8 (No BOM) as default.  Can you give me a specific example where this is known to break?  If OEM is needed, perhaps we could add a setting in the powershell json file but still default to utf8

iSazonov replied:
PR #3467#3467 haven't tests because (1) we did not have a full set of encoding tests in the time (2) we waited the encoding RFC (3) we only restore Windows PowerShell behavior by removing a breaking change.
About concrete examples. I speak about business applications. I have no idea how I can demonstrate it concrete. I understand that you hardly ever changed the OEM encoding on your computers. But ask yourself why are customers forced to do so? If MSTF wants to see it and understand it, it needs TAP with non-english enterprises. As I said earlier, the simple answer is that we have many business applications that only work with a specific OEM encoding and read/write files in OEM. PowerShell also works in _this environment_! The #3467 was caused by the fact that the scripts were no longer working and we needed to re-encode the files. The value of PowerShell Core disappears completely.

We can look at it on the other side. A year ago there was a categorical position that we should be backward compatible and that we should avoid making breaking changes as much as possible. The position was later changed so we want to facilitate the adaptation of UNIX users. This led to the creation of many incompatible changes. The current change from this series. These two principles have proved to be contradictory. We need to decide which one's in charge.

If the first means that we want to allow multiple Windows users to use PowerShell on Unix without much effort, and to get feedback to start the process of evolutionary convergence of platforms.

The second actually means that we're creating a brand new product that cannot be version 6, but must be PowerShell Core 1.0. It's fair to say that Windows users will have to migrate to a new product instead of using it directly. In this case, I agree that we can be ahead of the Windows 10 system and even its Notepad and use Utf8 by default like all other changes. Only then will it not be expected that the community will begin to use PowerShell Core extensively before version 3.
The second way is very dangerous. Microsoft can break the Windows PowerShell community and not build a new one-there are already well-known and popular alternatives (PHP, JS, Perl) that have proven their reliability and performance and have already formed a significant community. I've seen some examples of Microsoft already quickly destroying communities. The recovery took several years. It's sad. If it were my business, I would prefer to keep today's real customers, not future abstracts, and make the first step similar WSL like Windows PowerShell for Unix to capture a reliable bridgehead. Only then the community would be prepared to accept any breaking changes as understandable and necessary. Sorry for many words :-)

SteveL-MSFT replied:
It seems again that a viable option is to have a setting for powershell.runtimeconfig.json to have this be OEM for compat reasons.

mklement0 replied:
I haven't followed this closely, but what happened to the `$PSDefaultEncoding = 'WindowsLegacy'` proposal from the RFC?

mklement0 replied:
And, just to clarify: for backward compatibility, it's both OEM _and "ANSI"_ code pages that must be considered.

SteveL-MSFT replied:
@mklement0 the simplified version we scoped for 6.0.0 does not include `WindowsLegacy`

SteveL-MSFT replied:
@JamesWTruher can you open an issue to cover @iSazonov 's concern?

JamesWTruher replied:
@mklement0 WindowsLegacy was pulled waiting for customer need - opened https://github.com/PowerShell/PowerShell/issues/5204 to track.

dantraMSFT commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -29,13 +29,14 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
-        $CR  = [text.encoding]::unicode.GetBytes($asciiCR)
+        # create the expected - utf8 encoding without a bom
+        $encoding = [text.utf8encoding]::new($false)
+        $BOM = $encoding.GetPreamble()
```
This is very misleading; the test is verifying with a BOM yet the test is requesting the BOM.  If the Encoding class has a bug or was created incorrectly, the test will still pass.  In other words, if the array is empty, as expected, the test passes. If the array is non-empty, the test will still pass.

I suggest removing $BOM completely and not relying on the implicit empty array.

JamesWTruher replied:
done

dantraMSFT commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -29,13 +29,14 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
-        $TXT = [text.encoding]::unicode.GetBytes($asciiString)
-        $CR  = [text.encoding]::unicode.GetBytes($asciiCR)
+        # create the expected - utf8 encoding without a bom
+        $encoding = [text.utf8encoding]::new($false)
```
Suggest matching the casing of the type [Text.UTF8Encoding].

JamesWTruher replied:
I fixed up the cases of everything while I was there

dantraMSFT commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1:
```diff
@@ -99,13 +99,12 @@ baz
 
         $testPath = Join-Path -Path $TestDrive -ChildPath 'TailWithEncoding.txt'
-        $content | Set-Content -Path $testPath -Encoding $encoding
+        $content | Set-Content -Path $testPath -Encoding $encodingName
```
Should this test cover the ByteEncoding path?  It doesn't appear to.

JamesWTruher replied:
covered by the tests in engine/Basic

iSazonov commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -459,7 +478,7 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
             if (string.IsNullOrEmpty(encoding))
             {
-                // no parameter passed, default to Unicode (OS preferred)
+                // no parameter passed, default to  (OS preferred)
```
Typo?

JamesWTruher replied:
indeed so - fixed

iSazonov commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -445,10 +451,23 @@ internal static class EncodingConversion
+        // We will provide a partial list for tab completion we will convert to an encoding.
+        // If the user has an encoding object, we will cheerfully use it, but we can't use
+        // it as a validation set, as a user may have an specialty encoding which we have
+        // no idea about.
```
I think we could use `ArgumentCompletionsAttribute` #4835

iSazonov commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -500,9 +528,8 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
-                new string[] { Unknown, String, Unicode, BigEndianUnicode, Ascii, Utf8, Utf7, Utf32, Default, OEM });
-            string msg = StringUtil.Format(PathUtilsStrings.OutFile_WriteToFileEncodingUnknown,
-                encoding, validEncodingValues);
+                new string[] { Unknown, String, Unicode, BigEndianUnicode, Byte, Ascii, Utf8, Utf8Bom, Utf8NoBom, Utf7, Utf32, Default, OEM });
```
Can we reuse `TabCompletionResults` ?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +206,27 @@ public SwitchParameter NoClobber
-        [ValidateSetAttribute(new string[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; }
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
```
I think we could use `ArgumentCompletionsAttribute` #4835

JamesWTruher committed:
Update to follow coding style guidelines

JamesWTruher committed:
Fix cmdlet casing to follow guidelines

lzybkr commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -470,6 +489,9 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
                 return System.Text.Encoding.Unicode;
 
+            if (string.Equals(encoding, Byte, StringComparison.OrdinalIgnoreCase))
+                return byteEncoding;
```
If we aren't using a dictionary for fast lookup, we can at least order the tests from most likely (guessing of course) to least likely.

e.g., is Unknown really the most likely case? I hope not.

JamesWTruher replied:
reordered

lzybkr commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -515,5 +542,104 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
+            // the encoding for this is not bigendian and not BOM
+            public Encoding ActualEncoding = new UnicodeEncoding(false, false);
+            public override unsafe int GetByteCount(char* c, int i1) { throw new NotImplementedException(); }
+            public override int GetByteCount(char[] c, int i1, int i2) { throw new NotImplementedException(); }
```
I believe it's not hard for someone to get an instance of this encoding, so maybe we should be implementing all of these methods, assuming it's straightforward to do so (and mostly it seems like it is.)

JamesWTruher replied:
done

JamesWTruher committed:
Change the encoding to utf8 when the converter gets an empty string

refactor encoding class and other encoding utilities into new file
update redirection operator tests to
convert string matching for encoding into dictionary, which should make it more efficient

mklement0 commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -445,10 +451,23 @@ internal static class EncodingConversion
         internal const string Utf32 = "utf32";
         internal const string Default = "default";
         internal const string OEM = "oem";
+        internal const string Byte = "byte"; // This will return unicode, but 'byte' is special as it uses unicode encoding without a BOM
```
Not knowing the inner workings: How will this return _Unicode_? With `Get-Content`, you get a _byte_ array, and on output  (`Set-Content` / `Add-Content`) you're forced to provide a byte array _as input_ too (however, you constructed that).

JamesWTruher committed:
Change Encoding to remove byte as a possible encoding value

Add a new parameter to the filesystem provider cmdlets to indicate whether we want a byte stream
Unify Send-MailMessage with other cmdlets

JamesWTruher commented:
@adityapatwardhan @dantraMSFT @SteveL-MSFT I've believe I've made updates that address your blocking issues

SteveL-MSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7477,73 +7477,6 @@ public static Hashtable Invoke(System.Management.Automation.PowerShell ps, FileS
-    /// Defines the values that can be supplied as the encoding parameter in the
-    /// FileSystemContentDynamicParametersBase class.
-    /// </summary>
-    public enum FileSystemCmdletProviderEncoding
```
Jim, can you check if this API is in PowerShell Standard 3.0?  This enum isn't used at all now, right?  So leaving it in with Obsolete doesn't help.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and has agreed on:

1. name the parameter `-AsByteArray`
2. output warning if `-Encoding` is used with `-AsByteArray` that Encoding is not being used

mklement0 commented:
A message from the COSP (Crying Over Spilt Milk) department:

> output warning if -Encoding is used with -AsByteArray that Encoding is not being used

This results in an awkward user experience, because users will expect incompatible parameters to be in different parameter sets, with attempts to combine them resulting in an _error_, not a warning.

A step backward in usability, and a breaking change to boot - and no gain that I can discern.

I also agree with @iSazonov that forgoing compatibility with the legacy encoding behavior on Windows will dismay many Windows users, especially those whose "ANSI" / OEM code pages aren't based on the Latin alphabet.

If the idea is to focus on the Unix crowd first, I see trouble as well: broken quoting, broken CLI, lack of `&&` and `||`, ...

JamesWTruher committed:
Changed -Byte parameter to -AsByteStream

Created a warning when both -AsByteStream and -Encoding are used together
updated tests to check that the warning is emitted

SteveL-MSFT commented:
@mklement one of the considerations the committee acknowledged is the complexity of using multiple parametersets to make parameters mutually exclusive so we were ok with using a warning as a simplification until we can revisit the recalled RFC on making it easier to both define as well as understand mutually exclusive parameters

adityapatwardhan reviewed (approved):
LGTM

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +206,9 @@ public SwitchParameter NoClobber
-        public string Encoding { get; set; }
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
This is a public field. I think we need it to be a property.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -571,8 +572,9 @@ public SwitchParameter UseCulture
-        public string Encoding { get; set; }
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
Same here, should make it a property.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -46,14 +46,9 @@ public sealed class FormatHex : PSCmdlet
-        public string Encoding { get; set; } = "Ascii";
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
Same here.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/out-file/Out-File.cs:
```diff
@@ -72,25 +73,9 @@ public string LiteralPath
         /// </summary>
         ///
         [Parameter(Position = 1)]
-        [ValidateNotNullOrEmpty]
```
I think we still need this attribute, so that user cannot pass in a null value.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +206,9 @@ public SwitchParameter NoClobber
         [Parameter()]
-        [ValidateSetAttribute(new string[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; }
+        [ArgumentToEncodingTransformationAttribute()]
```
I think we need to add `AttributeNotNullOrEmpty` attribute. Before this change, $null value would be rejected because the validate attribute will fail. After this change, user is able to pass in a null value.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -571,8 +572,9 @@ public SwitchParameter UseCulture
-        [ValidateSetAttribute(new[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; }
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
```
We need a `ValidateNotNullOrEmtpy` attribute here.

JamesWTruher replied:
fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/format-hex/Format-Hex.cs:
```diff
@@ -46,14 +46,9 @@ public sealed class FormatHex : PSCmdlet
-            EncodingConversion.Ascii})]
-        public string Encoding { get; set; } = "Ascii";
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
```
Need `ValidateNullOrEmpty` here.

JamesWTruher replied:
fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs:
```diff
@@ -77,19 +77,9 @@ public SwitchParameter Force
-        private Encoding _encoding = System.Text.Encoding.UTF8;
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
Need to be a property.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs:
```diff
@@ -77,19 +77,9 @@ public SwitchParameter Force
-            }
-        }
-        private Encoding _encoding = System.Text.Encoding.UTF8;
+        [ArgumentToEncodingTransformationAttribute()]
```
Need `ValidateNullOrEmpty`.

JamesWTruher replied:
fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1201,19 +1202,9 @@ public SwitchParameter AllMatches
-        private System.Text.Encoding _textEncoding;
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
Same comments here.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs:
```diff
@@ -89,20 +89,14 @@ public SwitchParameter BodyAsHtml
-        private Encoding _encoding = new ASCIIEncoding();
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        [ArgumentToEncodingTransformationAttribute()]
+        public Encoding Encoding { get; set; } = new ASCIIEncoding();
```
It's better to use `Encoding.ASCII`, so that we don't create a new object.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -108,8 +108,9 @@ public SwitchParameter NoClobber
-        public string Encoding { get; set; } = "Unicode";
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
Should be a property.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs:
```diff
@@ -108,8 +108,9 @@ public SwitchParameter NoClobber
-        [ValidateSetAttribute(new string[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; } = "Unicode";
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
```
need `validatenotnulloremtpy`

JamesWTruher replied:
fixed

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1210,77 +1211,26 @@ internal static FileSystemCmdletProviderEncoding GetEncoding(string path)
             if (initialBytesAsAscii.IndexOfAny(nonPrintableCharacters) >= 0)
             {
-                return FileSystemCmdletProviderEncoding.Byte;
+                return Encoding.Unicode;
```
Returning `Encoding.Unicode` doesn't seem right. It looks we should deal with bytes in this case.

JamesWTruher replied:
There is no byte encoding, and FileSystemCmdletProviderEncoding.Byte wound up being converted to a unicode encoding in GetEncodingFromEnum (now removed) so there's actually no change here.

daxian-dbw reviewed (changes requested):
I'm now looking at `FileSystemProvider.cs`, but want to share my feedback early.

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1201,19 +1202,9 @@ public SwitchParameter AllMatches
-        private System.Text.Encoding _textEncoding;
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
+        public Encoding Encoding = ClrFacade.GetDefaultEncoding();
```
You have a few cases where you declare the parameter as a field instead of a property and it doesn't appear that you enforce non-null.  In this case, the null Encoding is passed to new StreamReader which throws an ArgumentNullException.   

In this particular case, the parameter is not marked as mandatory and an exception should not be raised if not provided.

JamesWTruher replied:
I've added `[ValidateNotNullOrEmpty]` to all of these.

JamesWTruher replied:
I've fixed up all the parameters to use [ValidateNotNullOrEmpty]

daxian-dbw commented on src/System.Management.Automation/utils/EncodingUtils.cs:
```diff
@@ -0,0 +1,136 @@
+        public override object Transform(EngineIntrinsics engineIntrinsics, object inputData)
+        {
+            string encodingName;
+            if (LanguagePrimitives.TryConvertTo<string>(inputData, out encodingName))
```
If the `inputData` is already an Encoding object, shouldn't we just let it pass by?

daxian-dbw replied:
And do we really need to convert `inputData` to `string` here? Why not just use `string encodingName = inputData as string`?

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7647,7 +7590,28 @@ public class FileSystemContentDynamicParametersBase
         [Parameter]
-        public FileSystemCmdletProviderEncoding Encoding { get; set; } = FileSystemCmdletProviderEncoding.String;
+        [ArgumentToEncodingTransformationAttribute()]
+        [ArgumentCompleter(typeof(EncodingArgumentCompleter))]
```
I think we also need a `ValidateNullOrEmpty` here.

JamesWTruher replied:
fixed

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1434,7 +1415,7 @@ private bool ProcessFile(string filename)
                 using (FileStream fs = new FileStream(filename, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
                 {
-                    using (StreamReader sr = new StreamReader(fs, _textEncoding))
+                    using (StreamReader sr = new StreamReader(fs, Encoding))
```
There's an ArgumentNullException path here if Encoding is null.

JamesWTruher replied:
fixed with `[ValidateNotNullOrEmpty]` attribute

JamesWTruher replied:
fixed with attribute

daxian-dbw commented on src/System.Management.Automation/resources/FileSystemProviderStrings.resx:
```diff
@@ -243,6 +243,9 @@
     <value>Cannot process path '{0}' because the target represents a reserved device name.</value>
   </data>
+  <data name="EncodingNotUsed" xml:space="preserve">
+    <value>Encoding not used when -AsByteStream specified.</value>
```
Minor comment: maybe change to `... when '-AsByteStream' is specified`?

JamesWTruher replied:
done

dantraMSFT reviewed (dismissed):
The change from property to field for some of the Encoding parameters needs to be reviewed for null references. MatchString.cs is one case where the parameter is a field, it is not mandatory, not marked as ValidateNotNull and is passed to an underlying StreamReader unchecked.

daxian-dbw commented on src/System.Management.Automation/utils/EncodingUtils.cs:
```diff
@@ -0,0 +1,136 @@
+                {
+                    encodings.Add(new CompletionResult(encoding, encoding, CompletionResultType.Text, encoding));
+                }
+                else if (encoding.IndexOf(wordToComplete, StringComparison.InvariantCultureIgnoreCase) == 0)
```
I think `wordToComplete` could have wildcards in it. For example, `Get-Content -Encoding *ianUnicode<tab>` ==> `Get-Content -Encoding BigEndianUnicode`

JamesWTruher replied:
I can add that

daxian-dbw commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -188,6 +188,12 @@ internal static void ReportFileOpenFailure(Cmdlet cmdlet, string filePath, Excep
+            return new StreamReader(fileStream, encoding);
+        }
+
         internal static StreamReader OpenStreamReader(PSCmdlet command, string filePath, string encoding, bool isLiteralPath)
```
Is this method still used anywhere? It looks to me not used anymore. If that's the case, then it's better to remove it.

JamesWTruher replied:
removed

JamesWTruher commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -6614,12 +6614,17 @@ public IContentReader GetContentReader(string path)
 
+                    if (usingByteEncoding && streamTypeSpecified)
+                    {
+                        WriteWarning(StringUtil.Format(FileSystemProviderStrings.EncodingNotUsed));
```
fixed

JamesWTruher commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -6765,12 +6770,17 @@ public IContentWriter GetContentWriter(string path)
 
+                    if (usingByteEncoding && streamTypeSpecified)
+                    {
+                        WriteWarning(StringUtil.Format(FileSystemProviderStrings.EncodingNotUsed));
```
fixed

JamesWTruher commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7658,39 +7622,10 @@ public class FileSystemContentDynamicParametersBase
-                return (Encoding != FileSystemCmdletProviderEncoding.String);
-            } // get
-        } // WasStreamTypeSpecified
+        public bool WasStreamTypeSpecified { get; set; }
```
done

JamesWTruher committed:
put ValidateNotNullOrEmpty attribute on Encoding parameters

Also improve encoding completion to handle wildcards
Also make sure that Encoding parameters are properties rather than fields.
Remove unused OpenStreamReader which takes a string since it's no longer in use

daxian-dbw commented on src/System.Management.Automation/utils/EncodingUtils.cs:
```diff
@@ -0,0 +1,139 @@
+            if (EncodingConversion.encodingMap.TryGetValue(encodingName, out foundEncoding))
+            {
+                return foundEncoding;
+            }
```
It looks to me we only care about convert a stirng to an encoding. I suggest change the body to the following
```
string encodingName = inputData as String;
Encoding foundEncoding;
if (encodingName != null && EncodingConversion.encodingMap.TryGetValue(encodingName, out foundEncoding))
{
    return foundEncoding;
}
return inputData.
```

JamesWTruher replied:
done

daxian-dbw commented on src/System.Management.Automation/utils/EncodingUtils.cs:
```diff
@@ -0,0 +1,139 @@
+
+    }
+
+    internal class EncodingArgumentCompleter : IArgumentCompleter
```
This is not necessary. You should use [`ArgumentCompletionsAttribute`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs#L172) instead for the parameters. Here is an example of using it: https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs#L210

JamesWTruher replied:
fixed

JamesWTruher committed:
Change from IArgumentCompleter attribute to ArgumentCompletions attribute for encoding attribute.

Simplify logic for converting string to encoding

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -212,8 +212,20 @@ public SwitchParameter NoClobber
+            EncodingConversion.Utf8Bom,
+            EncodingConversion.Utf8NoBom,
+            EncodingConversion.Utf32
+            )]
```
How about use `ArgumentCompletions(EncodingConversion.TabCompletionResults)` here? It would be easy for future updates too.

daxian-dbw replied:
Same comments for other instances of `ArgumentCompletions` usage.

JamesWTruher replied:
that doesn't work, because arrays aren't constant

daxian-dbw replied:
OK, then we are good here. #Closed

daxian-dbw commented:
@dantraMSFT Can you please take another look and update your review?

daxian-dbw commented:
I'm getting this PR merged so that it can be validated by today's daily builds.
Actually, this PR should have the last commit marked with '[Feature]' tag to trigger full build run.
